runtime/runtime-params-estimator/src/least_squares.rs

```diff
@@ -37,6 +40,77 @@ pub(crate) fn least_squares_method(
     (a, b, errs)
 }
 
+pub(crate) fn time_measurement_least_squares(
+    xs: &[u64],
+    ys: &[Ratio<u64>],
+) -> ((Ratio<u64>, Ratio<u64>), (Ratio<u64>, Ratio<u64>)) {
+    let t = least_squares_method(
+        xs,
+        &ys.iter().map(|ns| (ns * GAS_IN_NS).to_integer()).collect::<Vec<_>>(),
+    );
+    let (pos_t_base, neg_t_base) = split_pos_neg(t.0);
+    let (pos_t_factor, neg_t_factor) = split_pos_neg(t.1);
+
+    let neg_base = neg_t_base / GAS_IN_NS;
+    let neg_factor = neg_t_factor / GAS_IN_NS;
+    let pos_base = pos_t_base / GAS_IN_NS;
+    let pos_factor = pos_t_factor / GAS_IN_NS;
+
+    ((pos_base, pos_factor), (neg_base, neg_factor))
+}
+
+pub(crate) fn qemu_measurement_least_squares(
+    xs: &[u64],
+    ys: &[QemuMeasurement],
+) -> ((QemuMeasurement, QemuMeasurement), (QemuMeasurement, QemuMeasurement)) {
```

Comment:
Not too happy about these two, but also don't see an immediately better option, short-of rewriting the thing in Haskell and doing some cursed Function-fu

Reply:
I cannot disagree with that ^^ It was even worse before IMO, when they were cramped into the same function. But now the ugliness becomes more apparent, especially for the time-based method. Maybe I will find time to further improve it in another PR but this already feels like walking in the right direction to me.